Premier League 2025/26 Analytics Project
This project builds a football analytics database using Premier League 2025/26 player statistics from FBref.
--------------------------------------------------------------------------------------------------------------
Tools Used:
Python
,Pandas
,SQLite
,SQL

Project Goals:
Build a relational football database
,Clean and trasform raw FBref data
,Perform SQL analysis



# Import Libraries
Import the python libraries required for database management and data analysis.

In [1]:
import pandas as pd
import sqlite3 
print("Done")

Done


# Database Creation
Create a SQLite database and establish a connection for storing Premier League data.

In [2]:
conn= sqlite3.connect("premier_league_2025_26.db")
print("Database created")

Database created


# Create Teams Table
Create a table to store Premier League clubs and their unique Team IDs.

In [3]:
cursor = conn.cursor()

cursor.execute("""
CREATE TABLE IF NOT EXISTS Teams (
    TeamID INTEGER PRIMARY KEY,
    TeamName TEXT NOT NULL
)
""")

conn.commit()

print("Teams table created!")

Teams table created!


In [4]:
pd.read_sql("""
SELECT name
FROM sqlite_master
WHERE type='table';
""", conn)

,name
0,Teams
1,Players
2,PlayerStats


# Insert Premier League Clubs
Insert all 20 Premier League Clubs into the Teams Table.


In [5]:
teams = [
    (1, "Arsenal"),
    (2, "Aston Villa"),
    (3, "Bournemouth"),
    (4, "Brentford"),
    (5, "Brighton"),
    (6, "Burnley"),
    (7, "Chelsea"),
    (8, "Crystal Palace"),
    (9, "Everton"),
    (10, "Fulham"),
    (11, "Leeds United"),
    (12, "Liverpool"),
    (13, "Manchester City"),
    (14, "Manchester United"),
    (15, "Newcastle United"),
    (16, "Nottingham Forest"),
    (17, "Sunderland"),
    (18, "Tottenham Hotspur"),
    (19, "West Ham United"),
    (20, "Wolverhampton Wanderers")
]

cursor.executemany(
    "INSERT OR IGNORE INTO Teams (TeamID, TeamName) VALUES (?, ?)",
    teams
)

conn.commit()

print("20 teams inserted!")

20 teams inserted!


In [6]:
pd.read_sql("SELECT * FROM Teams", conn)

,TeamID,TeamName
0,1,Arsenal
1,2,Aston Villa
2,3,Bournemouth
3,4,Brentford
4,5,Brighton
5,6,Burnley
6,7,Chelsea
7,8,Crystal Palace
8,9,Everton
9,10,Fulham


# Create Players Table 
Create a table to store player information including name, team, position, age and nationality.

In [7]:
cursor.execute("""
CREATE TABLE IF NOT EXISTS Players (
    PlayerID INTEGER PRIMARY KEY,
    PlayerName TEXT NOT NULL,
    TeamID INTEGER NOT NULL,
    Position TEXT,

    FOREIGN KEY (TeamID)
    REFERENCES Teams(TeamID)
)
""")

conn.commit()

print("Players table created!")

Players table created!


In [8]:
pd.read_sql("""
SELECT name
FROM sqlite_master
WHERE type='table';
""", conn)

,name
0,Teams
1,Players
2,PlayerStats


In [9]:
pd.read_sql("SELECT * FROM Teams", conn)

,TeamID,TeamName
0,1,Arsenal
1,2,Aston Villa
2,3,Bournemouth
3,4,Brentford
4,5,Brighton
5,6,Burnley
6,7,Chelsea
7,8,Crystal Palace
8,9,Everton
9,10,Fulham


# Create Playerstats Table
Create a table to store player performance statistics for the 2025/26 season.

In [10]:
cursor.execute("""
CREATE TABLE IF NOT EXISTS PlayerStats (
    PlayerID INTEGER PRIMARY KEY,
    MatchesPlayed INTEGER,
    MinutesPlayed INTEGER,
    Goals INTEGER,
    Assists INTEGER,
    YellowCards INTEGER,
    RedCards INTEGER,

    FOREIGN KEY (PlayerID)
    REFERENCES Players(PlayerID)
)
""")

conn.commit()

print("PlayerStats table created!")

PlayerStats table created!


In [11]:
pd.read_sql("""
SELECT name
FROM sqlite_master
WHERE type='table';
""", conn)

,name
0,Teams
1,Players
2,PlayerStats


In [12]:
pd.read_sql("""
PRAGMA table_info(Players);
""", conn)

,cid,name,type,notnull,dflt_value,pk
0,0,PlayerID,INTEGER,0,None,1
1,1,PlayerName,TEXT,1,None,0
2,2,TeamID,INTEGER,1,None,0
3,3,Position,TEXT,0,None,0
4,4,Age,INTEGER,0,None,0
5,5,Nationality,TEXT,0,None,0


In [13]:
cursor.execute("DROP TABLE IF EXISTS Players")
conn.commit()

print("Players table removed")

Players table removed


In [14]:
cursor.execute("""
CREATE TABLE Players (
    PlayerID INTEGER PRIMARY KEY,
    PlayerName TEXT NOT NULL,
    TeamID INTEGER NOT NULL,
    Position TEXT,
    Age INTEGER,
    Nationality TEXT,

    FOREIGN KEY (TeamID)
    REFERENCES Teams(TeamID)
)
""")

conn.commit()

print("New Players table created")

New Players table created


In [15]:
pd.read_sql("""
PRAGMA table_info(Players);
""", conn)

,cid,name,type,notnull,dflt_value,pk
0,0,PlayerID,INTEGER,0,None,1
1,1,PlayerName,TEXT,1,None,0
2,2,TeamID,INTEGER,1,None,0
3,3,Position,TEXT,0,None,0
4,4,Age,INTEGER,0,None,0
5,5,Nationality,TEXT,0,None,0


In [16]:
cursor.execute("DROP TABLE IF EXISTS PlayerStats")
conn.commit()

In [17]:
cursor.execute("""
CREATE TABLE PlayerStats (
    PlayerID INTEGER PRIMARY KEY,
    MatchesPlayed INTEGER,
    Starts INTEGER,
    MinutesPlayed INTEGER,
    Nineties REAL,
    Goals INTEGER,
    Assists INTEGER,
    GoalsAssists INTEGER,
    NonPenaltyGoals INTEGER,
    PenaltyGoals INTEGER,
    PenaltyAttempts INTEGER,
    YellowCards INTEGER,
    RedCards INTEGER,

    FOREIGN KEY (PlayerID)
    REFERENCES Players(PlayerID)
)
""")

conn.commit()

print("PlayerStats table created")

PlayerStats table created


In [18]:
import requests
import pandas as pd

url = "https://fbref.com/en/comps/9/stats/Premier-League-Stats"

headers = {
    "User-Agent": "Mozilla/5.0"
}

response = requests.get(url, headers=headers)

print(response.status_code)

403


In [19]:
pd.read_sql("""
PRAGMA table_info(PlayerStats);
""", conn)

,cid,name,type,notnull,dflt_value,pk
0,0,PlayerID,INTEGER,0,None,1
1,1,MatchesPlayed,INTEGER,0,None,0
2,2,Starts,INTEGER,0,None,0
3,3,MinutesPlayed,INTEGER,0,None,0
4,4,Nineties,REAL,0,None,0
5,5,Goals,INTEGER,0,None,0
6,6,Assists,INTEGER,0,None,0
7,7,GoalsAssists,INTEGER,0,None,0
8,8,NonPenaltyGoals,INTEGER,0,None,0
9,9,PenaltyGoals,INTEGER,0,None,0


In [20]:
import os

print(os.getcwd())

C:\Users\Lenovo\PremierLeague2025_26_Project


In [21]:
import os

print(os.getcwd())

C:\Users\Lenovo\PremierLeague2025_26_Project


# Import FBref Dataset
Load the exported FBref dataset into a Pandas DataFrame for cleaning and transformation.

In [22]:
import pandas as pd

df = pd.read_excel("data/player_stats.xlsx")

print(df.shape)

(633, 25)


In [23]:
print(df.columns.tolist())

['Rk', 'Player', 'Nation', 'Pos', 'Squad', 'Age', 'Born', 'MP', 'Starts', 'Min', '90s', 'Gls', 'Ast', 'G+A', 'G-PK', 'PK', 'PKatt', 'CrdY', 'CrdR', 'Gls.1', 'Ast.1', 'G+A.1', 'G-PK.1', 'G+A-PK', 'Matches']


In [24]:
df.head()

,Rk,Player,Nation,Pos,Squad,Age,Born,MP,Starts,Min,...,PK,PKatt,CrdY,CrdR,Gls.1,Ast.1,G+A.1,G-PK.1,G+A-PK,Matches
0,▲,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,1,Brenden Aaronson,us USA,"MF,FW",Leeds United,24,2000,37,30,2449,...,0,0,3,0,0.15,0.18,0.33,0.15,0.33,Matches
2,2,Jerome Abbey,eng ENG,MF,Wolves,15,2009,1,0,17,...,0,0,0,0,0,0,0,0,0,Matches
3,3,Zach Abbott,eng ENG,DF,Nottingham Forest,19,2006,3,2,164,...,0,0,0,0,0,0,0,0,0,Matches
4,4,Tammy Abraham,eng ENG,FW,Aston Villa,27,1997,12,2,286,...,0,0,0,0,0.63,0,0.63,0.63,0.63,Matches


In [25]:
df = df[df["Player"] != "Player"]

print(df.shape)

(606, 25)


# Data Cleaning
Remove duplicate headers, empty rows and other inconsistencies from the imported dataset.

In [26]:
df = df.dropna(how="all")

print(df.shape)

(606, 25)


In [27]:
print(df["Player"].nunique())

537


In [28]:
df.head()

,Rk,Player,Nation,Pos,Squad,Age,Born,MP,Starts,Min,...,PK,PKatt,CrdY,CrdR,Gls.1,Ast.1,G+A.1,G-PK.1,G+A-PK,Matches
0,▲,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,1,Brenden Aaronson,us USA,"MF,FW",Leeds United,24,2000,37,30,2449,...,0,0,3,0,0.15,0.18,0.33,0.15,0.33,Matches
2,2,Jerome Abbey,eng ENG,MF,Wolves,15,2009,1,0,17,...,0,0,0,0,0,0,0,0,0,Matches
3,3,Zach Abbott,eng ENG,DF,Nottingham Forest,19,2006,3,2,164,...,0,0,0,0,0,0,0,0,0,Matches
4,4,Tammy Abraham,eng ENG,FW,Aston Villa,27,1997,12,2,286,...,0,0,0,0,0.63,0,0.63,0.63,0.63,Matches


In [29]:
df = df[df["Player"] != "Player"]

df = df.dropna(how="all")

print(df.shape)

print(df["Player"].nunique())

(606, 25)
537


In [30]:
df[df["Player"].isna()]

,Rk,Player,Nation,Pos,Squad,Age,Born,MP,Starts,Min,...,PK,PKatt,CrdY,CrdR,Gls.1,Ast.1,G+A.1,G-PK.1,G+A-PK,Matches
0,▲,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
21,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Playing Time,NaN,NaN,...,NaN,NaN,NaN,NaN,Per 90 Minutes,NaN,NaN,NaN,NaN,NaN
23,▲,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
44,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Playing Time,NaN,NaN,...,NaN,NaN,NaN,NaN,Per 90 Minutes,NaN,NaN,NaN,NaN,NaN
46,▲,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
67,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Playing Time,NaN,NaN,...,NaN,NaN,NaN,NaN,Per 90 Minutes,NaN,NaN,NaN,NaN,NaN
69,▲,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
90,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Playing Time,NaN,NaN,...,NaN,NaN,NaN,NaN,Per 90 Minutes,NaN,NaN,NaN,NaN,NaN
92,▲,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
113,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Playing Time,NaN,NaN,...,NaN,NaN,NaN,NaN,Per 90 Minutes,NaN,NaN,NaN,NaN,NaN


In [31]:
df["Player"].value_counts().head(20)

Player
Tyrique George          2
Facundo Buonanotte      2
Oscar Bobb              2
Marc Guiu               2
Brennan Johnson         2
Marc Guéhi              2
Eberechi Eze            2
Harvey Elliott          2
Evann Guessand          2
Douglas Luiz            2
Antoine Semenyo         2
Jørgen Strand Larsen    2
Adama Traoré            2
James Ward-Prowse       2
Zach Abbott             1
Tammy Abraham           1
Joshua Acheampong       1
Tyler Adams             1
André                   1
Nayef Aguerd            1
Name: count, dtype: int64

In [32]:
df["Squad"].nunique()

20

In [33]:
df = df.dropna(subset=["Player"])

print(df.shape)
print(df["Player"].nunique())

(551, 25)
537


In [34]:
df["Squad"].nunique()

20

In [35]:
df["Squad"].sort_values().unique()

array(['Arsenal', 'Aston Villa', 'Bournemouth', 'Brentford', 'Brighton',
       'Burnley', 'Chelsea', 'Crystal Palace', 'Everton', 'Fulham',
       'Leeds United', 'Liverpool', 'Manchester City', 'Manchester Utd',
       'Newcastle United', 'Nottingham Forest', 'Sunderland',
       'Tottenham Hotspur', 'West Ham United', 'Wolves'], dtype=object)

# Team Mapping
Match FBref squad names to the Team IDs stored in the database.

In [36]:
team_mapping = {
    "Arsenal": 1,
    "Aston Villa": 2,
    "Bournemouth": 3,
    "Brentford": 4,
    "Brighton": 5,
    "Burnley": 6,
    "Chelsea": 7,
    "Crystal Palace": 8,
    "Everton": 9,
    "Fulham": 10,
    "Leeds United": 11,
    "Liverpool": 12,
    "Manchester City": 13,
    "Manchester Utd": 14,
    "Newcastle United": 15,
    "Nottingham Forest": 16,
    "Sunderland": 17,
    "Tottenham Hotspur": 18,
    "West Ham United": 19,
    "Wolves": 20
}

In [37]:
df["TeamID"]=df["Squad"].map(team_mapping)
df[["Player","Squad","TeamID"]].head()

,Player,Squad,TeamID
1,Brenden Aaronson,Leeds United,11
2,Jerome Abbey,Wolves,20
3,Zach Abbott,Nottingham Forest,16
4,Tammy Abraham,Aston Villa,2
5,Joshua Acheampong,Chelsea,7


In [38]:
print(df["TeamID"].isna().sum())

0


In [39]:
df.head()

,Rk,Player,Nation,Pos,Squad,Age,Born,MP,Starts,Min,...,PKatt,CrdY,CrdR,Gls.1,Ast.1,G+A.1,G-PK.1,G+A-PK,Matches,TeamID
1,1,Brenden Aaronson,us USA,"MF,FW",Leeds United,24,2000,37,30,2449,...,0,3,0,0.15,0.18,0.33,0.15,0.33,Matches,11
2,2,Jerome Abbey,eng ENG,MF,Wolves,15,2009,1,0,17,...,0,0,0,0,0,0,0,0,Matches,20
3,3,Zach Abbott,eng ENG,DF,Nottingham Forest,19,2006,3,2,164,...,0,0,0,0,0,0,0,0,Matches,16
4,4,Tammy Abraham,eng ENG,FW,Aston Villa,27,1997,12,2,286,...,0,0,0,0.63,0,0.63,0.63,0.63,Matches,2
5,5,Joshua Acheampong,eng ENG,DF,Chelsea,19,2006,17,8,666,...,0,1,0,0.14,0,0.14,0.14,0.14,Matches,7


In [40]:
df[["Player","Nation","Pos","Squad","Age","TeamID"]].head(10)

,Player,Nation,Pos,Squad,Age,TeamID
1,Brenden Aaronson,us USA,"MF,FW",Leeds United,24,11
2,Jerome Abbey,eng ENG,MF,Wolves,15,20
3,Zach Abbott,eng ENG,DF,Nottingham Forest,19,16
4,Tammy Abraham,eng ENG,FW,Aston Villa,27,2
5,Joshua Acheampong,eng ENG,DF,Chelsea,19,7
6,Tyler Adams,us USA,MF,Bournemouth,26,3
7,Tosin Adarabioyo,eng ENG,DF,Chelsea,27,7
8,Simon Adingra,ci CIV,"MF,FW",Sunderland,23,17
9,Amine Adli,ma MAR,MF,Bournemouth,25,3
10,Emmanuel Agbadou,ci CIV,DF,Wolves,28,20


# Create Player DataFrame
Extract player information and prepare it for insertion into Players table

In [41]:
players_df = df[[
    "Player",
    "TeamID",
    "Pos",
    "Age",
    "Nation"
]].copy()

players_df.head()

,Player,TeamID,Pos,Age,Nation
1,Brenden Aaronson,11,"MF,FW",24,us USA
2,Jerome Abbey,20,MF,15,eng ENG
3,Zach Abbott,16,DF,19,eng ENG
4,Tammy Abraham,2,FW,27,eng ENG
5,Joshua Acheampong,7,DF,19,eng ENG


In [42]:
players_df = players_df.rename(columns={
    "Player": "PlayerName",
    "Pos": "Position",
    "Nation": "Nationality"
})

players_df.head()

,PlayerName,TeamID,Position,Age,Nationality
1,Brenden Aaronson,11,"MF,FW",24,us USA
2,Jerome Abbey,20,MF,15,eng ENG
3,Zach Abbott,16,DF,19,eng ENG
4,Tammy Abraham,2,FW,27,eng ENG
5,Joshua Acheampong,7,DF,19,eng ENG


# Generate Player IDs
Create a unique PlayerID for every player record

In [43]:
players_df.insert(
    0,
    "PlayerID",
    range(1, len(players_df) + 1)
)

players_df.head()

,PlayerID,PlayerName,TeamID,Position,Age,Nationality
1,1,Brenden Aaronson,11,"MF,FW",24,us USA
2,2,Jerome Abbey,20,MF,15,eng ENG
3,3,Zach Abbott,16,DF,19,eng ENG
4,4,Tammy Abraham,2,FW,27,eng ENG
5,5,Joshua Acheampong,7,DF,19,eng ENG


In [44]:
players_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 551 entries, 1 to 632
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   PlayerID     551 non-null    int64 
 1   PlayerName   551 non-null    object
 2   TeamID       551 non-null    int64 
 3   Position     551 non-null    object
 4   Age          551 non-null    object
 5   Nationality  551 non-null    object
dtypes: int64(2), object(4)
memory usage: 30.1+ KB


In [45]:
players_df["Age"].head(20)

1     24
2     15
3     19
4     27
5     19
6     26
7     27
8     23
9     25
10    28
11    29
12    28
13    24
14    27
15    30
16    22
17    28
18    32
19    20
20    20
Name: Age, dtype: object

In [46]:
players_df["Age"] = players_df["Age"].astype(int)

In [47]:
players_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 551 entries, 1 to 632
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   PlayerID     551 non-null    int64 
 1   PlayerName   551 non-null    object
 2   TeamID       551 non-null    int64 
 3   Position     551 non-null    object
 4   Age          551 non-null    int64 
 5   Nationality  551 non-null    object
dtypes: int64(3), object(3)
memory usage: 30.1+ KB


In [48]:
pd.read_sql_query(
    "PRAGMA table_info(Players)",
    conn
)

,cid,name,type,notnull,dflt_value,pk
0,0,PlayerID,INTEGER,0,None,1
1,1,PlayerName,TEXT,1,None,0
2,2,TeamID,INTEGER,1,None,0
3,3,Position,TEXT,0,None,0
4,4,Age,INTEGER,0,None,0
5,5,Nationality,TEXT,0,None,0


# Insert Players into Database
Load 551 player records into the Players table.

In [49]:
players_df.to_sql(
    "Players",
    conn,
    if_exists="append",
    index=False
)

551

In [50]:
conn.commit()

In [51]:
pd.read_sql_query(
    "SELECT COUNT(*) AS TotalPlayers FROM Players",
    conn
)

,TotalPlayers
0,551


In [52]:
pd.read_sql_query(
    "PRAGMA table_info(PlayerStats)",
    conn
)

,cid,name,type,notnull,dflt_value,pk
0,0,PlayerID,INTEGER,0,None,1
1,1,MatchesPlayed,INTEGER,0,None,0
2,2,Starts,INTEGER,0,None,0
3,3,MinutesPlayed,INTEGER,0,None,0
4,4,Nineties,REAL,0,None,0
5,5,Goals,INTEGER,0,None,0
6,6,Assists,INTEGER,0,None,0
7,7,GoalsAssists,INTEGER,0,None,0
8,8,NonPenaltyGoals,INTEGER,0,None,0
9,9,PenaltyGoals,INTEGER,0,None,0


# Create Player statistics DataFrame
Extract performance statistics and prepare them for insertion into the PlayerStats table.

In [53]:
playerstats_df = pd.DataFrame({
    "PlayerID": players_df["PlayerID"],
    "MatchesPlayed": df["MP"],
    "Starts": df["Starts"],
    "MinutesPlayed": df["Min"],
    "Nineties": df["90s"],
    "Goals": df["Gls"],
    "Assists": df["Ast"],
    "GoalsAssists": df["G+A"],
    "NonPenaltyGoals": df["G-PK"],
    "PenaltyGoals": df["PK"],
    "PenaltyAttempts": df["PKatt"],
    "YellowCards": df["CrdY"],
    "RedCards": df["CrdR"]
})

playerstats_df.head()

,PlayerID,MatchesPlayed,Starts,MinutesPlayed,Nineties,Goals,Assists,GoalsAssists,NonPenaltyGoals,PenaltyGoals,PenaltyAttempts,YellowCards,RedCards
1,1,37,30,2449,27.2,4,5,9,4,0,0,3,0
2,2,1,0,17,0.2,0,0,0,0,0,0,0,0
3,3,3,2,164,1.8,0,0,0,0,0,0,0,0
4,4,12,2,286,3.2,2,0,2,2,0,0,0,0
5,5,17,8,666,7.4,1,0,1,1,0,0,1,0


In [54]:
playerstats_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 551 entries, 1 to 632
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   PlayerID         551 non-null    int64 
 1   MatchesPlayed    551 non-null    object
 2   Starts           551 non-null    object
 3   MinutesPlayed    551 non-null    object
 4   Nineties         551 non-null    object
 5   Goals            551 non-null    object
 6   Assists          551 non-null    object
 7   GoalsAssists     551 non-null    object
 8   NonPenaltyGoals  551 non-null    object
 9   PenaltyGoals     551 non-null    object
 10  PenaltyAttempts  551 non-null    object
 11  YellowCards      551 non-null    object
 12  RedCards         551 non-null    object
dtypes: int64(1), object(12)
memory usage: 60.3+ KB


# Data Type Conversion
Convert statistical columns from text format to numeric format before database insertion.

In [55]:
numeric_columns = [
    "MatchesPlayed",
    "Starts",
    "MinutesPlayed",
    "Nineties",
    "Goals",
    "Assists",
    "GoalsAssists",
    "NonPenaltyGoals",
    "PenaltyGoals",
    "PenaltyAttempts",
    "YellowCards",
    "RedCards"
]

for col in numeric_columns:
    playerstats_df[col] = pd.to_numeric(playerstats_df[col])

In [56]:
playerstats_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 551 entries, 1 to 632
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   PlayerID         551 non-null    int64  
 1   MatchesPlayed    551 non-null    int64  
 2   Starts           551 non-null    int64  
 3   MinutesPlayed    551 non-null    int64  
 4   Nineties         551 non-null    float64
 5   Goals            551 non-null    int64  
 6   Assists          551 non-null    int64  
 7   GoalsAssists     551 non-null    int64  
 8   NonPenaltyGoals  551 non-null    int64  
 9   PenaltyGoals     551 non-null    int64  
 10  PenaltyAttempts  551 non-null    int64  
 11  YellowCards      551 non-null    int64  
 12  RedCards         551 non-null    int64  
dtypes: float64(1), int64(12)
memory usage: 60.3 KB


# Insert Player Statistics into DataBase
Load 551 player-statistic records into the PlayerStats Table.

In [57]:
playerstats_df.to_sql(
    "PlayerStats",
    conn,
    if_exists="append",
    index=False
)

551

In [58]:
conn.commit()

In [59]:
pd.read_sql_query(
    "SELECT COUNT(*) AS TotalStatsRecords FROM PlayerStats",
    conn
)

,TotalStatsRecords
0,551


# SQL analysis
Run sql quaries to analyse player performance and generate football insights.

# Query 1 - Top Goalscorers


In [60]:
top_goals_query = """
SELECT
    p.PlayerName,
    ps.Goals
FROM Players AS p
JOIN PlayerStats AS ps
    ON p.PlayerID = ps.PlayerID
ORDER BY ps.Goals DESC
LIMIT 10;
"""

In [61]:
top_goals_df = pd.read_sql_query(top_goals_query, conn)

top_goals_df

,PlayerName,Goals
0,Erling Haaland,27
1,Igor Thiago,22
2,Ollie Watkins,16
3,Morgan Gibbs-White,15
4,João Pedro,15
5,Dominic Calvert-Lewin,14
6,Viktor Gyökeres,14
7,Eli Junior Kroupi,13
8,Danny Welbeck,13
9,Jean-Philippe Mateta,12


# Query 2 - Top Assist Providers

In [62]:
top_assists_query = """
SELECT
    p.PlayerName,
    ps.Assists
FROM Players AS p
JOIN PlayerStats AS ps
    ON p.PlayerID = ps.PlayerID
ORDER BY ps.Assists DESC
LIMIT 10;
"""

In [63]:
top_assists_df = pd.read_sql_query(top_assists_query, conn)

top_assists_df

,PlayerName,Assists
0,Bruno Fernandes,21
1,Rayan Cherki,12
2,Jarrod Bowen,11
3,Erling Haaland,8
4,James Garner,7
5,Mohamed Salah,7
6,Dominik Szoboszlai,7
7,Harry Wilson,7
8,Lucas Digne,6
9,Jack Grealish,6


# Query 3 - Team Discipline Score

In [64]:
team_discipline_query = """
SELECT
    t.TeamName,
    SUM(ps.YellowCards) AS YellowCards,
    SUM(ps.RedCards) AS RedCards,
    SUM(ps.YellowCards + (ps.RedCards * 2)) AS DisciplineScore
FROM Teams AS t
JOIN Players AS p
    ON t.TeamID = p.TeamID
JOIN PlayerStats AS ps
    ON p.PlayerID = ps.PlayerID
GROUP BY t.TeamName
ORDER BY DisciplineScore DESC;
"""

In [65]:
team_discipline_df = pd.read_sql_query(
    team_discipline_query,
    conn
)

team_discipline_df

,TeamName,YellowCards,RedCards,DisciplineScore
0,Chelsea,98,8,114
1,Tottenham Hotspur,101,4,109
2,Bournemouth,88,2,92
3,Sunderland,82,3,88
4,Wolverhampton Wanderers,80,3,86
5,Brighton,86,0,86
6,Everton,74,4,82
7,Crystal Palace,76,2,80
8,Fulham,75,1,77
9,West Ham United,69,3,75


# Query 4 - Average age of each team

In [66]:
average_age_by_team_query = """
SELECT
    t.TeamName,
    ROUND(AVG(p.Age), 1) AS AverageAge
FROM Teams AS t
JOIN Players AS p
    ON t.TeamID = p.TeamID
GROUP BY t.TeamName
ORDER BY AverageAge DESC;
"""

In [67]:
average_age_by_team_df = pd.read_sql_query(
    average_age_by_team_query,
    conn
)

average_age_by_team_df

,TeamName,AverageAge
0,Fulham,26.8
1,Leeds United,26.1
2,Newcastle United,25.9
3,Aston Villa,25.9
4,Burnley,25.8
5,Everton,25.6
6,Liverpool,25.5
7,West Ham United,25.3
8,Nottingham Forest,25.3
9,Arsenal,25.0


# Query 5 - Total goals scored by each team


In [68]:
goals_by_team_query = """
SELECT
    t.TeamName,
    SUM(ps.Goals) AS TotalGoals
FROM Teams AS t
JOIN Players AS p
    ON t.TeamID = p.TeamID
JOIN PlayerStats AS ps
    ON p.PlayerID = ps.PlayerID
GROUP BY t.TeamName
ORDER BY TotalGoals DESC;
"""


In [69]:
goals_by_team_df = pd.read_sql_query(
    goals_by_team_query,
    conn
)

goals_by_team_df

,TeamName,TotalGoals
0,Manchester City,74
1,Arsenal,67
2,Manchester United,66
3,Liverpool,61
4,Chelsea,57
5,Bournemouth,57
6,Newcastle United,53
7,Brentford,53
8,Aston Villa,53
9,Brighton,51


# Query 6 - Players with the most goal contributions per 90 (G/A)

In [70]:
goal_contributions_per_90_query = """
SELECT
    p.PlayerName,
    ps.Goals,
    ps.Assists,
    ps.Nineties,
    ROUND((ps.Goals + ps.Assists) / ps.Nineties, 2) AS GoalContributionsPer90
FROM Players AS p
JOIN PlayerStats AS ps
    ON p.PlayerID = ps.PlayerID
WHERE ps.Nineties >= 10
ORDER BY GoalContributionsPer90 DESC
LIMIT 10;
"""

In [71]:
goal_contributions_per_90_df = pd.read_sql_query(
    goal_contributions_per_90_query,
    conn
)

goal_contributions_per_90_df

,PlayerName,Goals,Assists,Nineties,GoalContributionsPer90
0,Erling Haaland,27,8,32.8,1.07
1,Bruno Fernandes,9,21,34.0,0.88
2,Rayan Cherki,4,12,19.8,0.81
3,Hugo Ekitike,11,4,20.0,0.75
4,Eli Junior Kroupi,13,0,18.6,0.70
5,Richarlison,11,4,21.8,0.69
6,João Pedro,15,5,29.6,0.68
7,Benjamin Šeško,11,1,18.3,0.66
8,Antoine Semenyo,10,3,20.0,0.65
9,Igor Thiago,22,1,36.5,0.63


# Query 7 - Players with the highest percentage of team goals.

In [72]:
team_goals_query = """
SELECT
    p.TeamID,
    SUM(ps.Goals) AS TeamGoals
FROM Players AS p
JOIN PlayerStats AS ps
    ON p.PlayerID = ps.PlayerID
GROUP BY p.TeamID;
"""

In [73]:
team_goals_df = pd.read_sql_query(
    team_goals_query,
    conn
)

team_goals_df

,TeamID,TeamGoals
0,1,67
1,2,53
2,3,57
3,4,53
4,5,51
5,6,35
6,7,57
7,8,40
8,9,45
9,10,43


In [74]:
percentage_team_goals_query = """
SELECT
    p.PlayerName,
    t.TeamName,
    ps.Goals,
    tg.TeamGoals,
    ROUND((ps.Goals * 100.0) / tg.TeamGoals, 1) AS GoalPercentage
FROM Players AS p

JOIN PlayerStats AS ps
    ON p.PlayerID = ps.PlayerID

JOIN Teams AS t
    ON p.TeamID = t.TeamID

JOIN (
    SELECT
        p.TeamID,
        SUM(ps.Goals) AS TeamGoals
    FROM Players AS p
    JOIN PlayerStats AS ps
        ON p.PlayerID = ps.PlayerID
    GROUP BY p.TeamID
) AS tg
    ON p.TeamID = tg.TeamID

WHERE ps.Nineties >= 10

ORDER BY GoalPercentage DESC
LIMIT 10;
"""

In [75]:
percentage_team_goals_df = pd.read_sql_query(
    percentage_team_goals_query,
    conn
)

percentage_team_goals_df

,PlayerName,TeamName,Goals,TeamGoals,GoalPercentage
0,Igor Thiago,Brentford,22,53,41.5
1,Erling Haaland,Manchester City,27,74,36.5
2,Morgan Gibbs-White,Nottingham Forest,15,47,31.9
3,Zian Flemming,Burnley,11,35,31.4
4,Ollie Watkins,Aston Villa,16,53,30.2
5,Jean-Philippe Mateta,Crystal Palace,12,40,30.0
6,Dominic Calvert-Lewin,Leeds United,14,48,29.2
7,João Pedro,Chelsea,15,57,26.3
8,Danny Welbeck,Brighton,13,51,25.5
9,Richarlison,Tottenham Hotspur,11,47,23.4


# Query 8 - Total number of assists by each team

In [76]:
assists_by_team_query = """
SELECT
    t.TeamName,
    SUM(ps.Assists) AS TotalAssists
FROM Teams AS t
JOIN Players AS p
    ON t.TeamID = p.TeamID
JOIN PlayerStats AS ps
    ON p.PlayerID = ps.PlayerID
GROUP BY t.TeamName
ORDER BY TotalAssists DESC;
"""

In [77]:
assists_by_team_df = pd.read_sql_query(
    assists_by_team_query,
    conn
)

assists_by_team_df

,TeamName,TotalAssists
0,Manchester City,57
1,Arsenal,51
2,Manchester United,46
3,Liverpool,44
4,Aston Villa,40
5,Chelsea,39
6,Tottenham Hotspur,36
7,Bournemouth,35
8,Nottingham Forest,34
9,Everton,34


# Query 9 - Top Players with the most goal contributions

In [78]:
goal_contributions_query = """
SELECT
    p.PlayerName,
    ps.Goals,
    ps.Assists,
    (ps.Goals + ps.Assists) AS GoalContributions
FROM Players AS p
JOIN PlayerStats AS ps
    ON p.PlayerID = ps.PlayerID
ORDER BY GoalContributions DESC
LIMIT 10;
"""

In [79]:
goal_contributions_df = pd.read_sql_query(
    goal_contributions_query,
    conn
)

goal_contributions_df

,PlayerName,Goals,Assists,GoalContributions
0,Erling Haaland,27,8,35
1,Bruno Fernandes,9,21,30
2,Igor Thiago,22,1,23
3,Jarrod Bowen,9,11,20
4,João Pedro,15,5,20
5,Morgan Gibbs-White,15,4,19
6,Ollie Watkins,16,3,19
7,Harry Wilson,10,7,17
8,Rayan Cherki,4,12,16
9,Morgan Rogers,10,6,16


# Query 10 - Players with the highest Goals per 90

In [80]:
goals_per_90_query = """
SELECT
    p.PlayerName,
    ps.Goals,
    ps.Nineties,
    ROUND(ps.Goals / ps.Nineties, 2) AS GoalsPer90
FROM Players AS p
JOIN PlayerStats AS ps
    ON p.PlayerID = ps.PlayerID
WHERE ps.Nineties >= 10
ORDER BY GoalsPer90 DESC
LIMIT 10;
"""

In [81]:
goals_per_90_df = pd.read_sql_query(
    goals_per_90_query,
    conn
)

goals_per_90_df

,PlayerName,Goals,Nineties,GoalsPer90
0,Erling Haaland,27,32.8,0.82
1,Eli Junior Kroupi,13,18.6,0.70
2,Benjamin Šeško,11,18.3,0.60
3,Igor Thiago,22,36.5,0.60
4,Viktor Gyökeres,14,24.7,0.57
5,Zian Flemming,11,19.6,0.56
6,Hugo Ekitike,11,20.0,0.55
7,Beto,9,17.4,0.52
8,Danny Welbeck,13,25.1,0.52
9,João Pedro,15,29.6,0.51


# Query 11 - Players with the most minutes played

In [82]:
minutes_played_query = """
SELECT
    p.PlayerName,
    t.TeamName,
    ps.MinutesPlayed
FROM Players AS p
JOIN PlayerStats AS ps
    ON p.PlayerID = ps.PlayerID
JOIN Teams AS t
    ON p.TeamID = t.TeamID
ORDER BY ps.MinutesPlayed DESC
LIMIT 10;
"""

In [83]:
minutes_played_df = pd.read_sql_query(
    minutes_played_query,
    conn
)

minutes_played_df

,PlayerName,TeamName,MinutesPlayed
0,Bernd Leno,Fulham,3420
1,Đorđe Petrović,Bournemouth,3420
2,Jordan Pickford,Everton,3420
3,Virgil van Dijk,Liverpool,3420
4,Bart Verbruggen,Brighton,3420
5,James Garner,Everton,3413
6,Jarrod Bowen,West Ham United,3403
7,Adrien Truffert,Bournemouth,3379
8,Nikola Milenković,Nottingham Forest,3375
9,Elliot Anderson,Nottingham Forest,3331


# Query 12 - Top Goalscorers from each team

In [84]:
top_scorer_each_team_query = """
SELECT
    t.TeamName,
    p.PlayerName,
    ps.Goals
FROM Players AS p
JOIN PlayerStats AS ps
    ON p.PlayerID = ps.PlayerID
JOIN Teams AS t
    ON p.TeamID = t.TeamID
WHERE ps.Goals = (
    SELECT MAX(ps2.Goals)
    FROM Players AS p2
    JOIN PlayerStats AS ps2
        ON p2.PlayerID = ps2.PlayerID
    WHERE p2.TeamID = p.TeamID
)
ORDER BY t.TeamName;
"""

In [85]:
top_scorer_each_team_df = pd.read_sql_query(
    top_scorer_each_team_query,
    conn
)

top_scorer_each_team_df

,TeamName,PlayerName,Goals
0,Arsenal,Viktor Gyökeres,14
1,Aston Villa,Ollie Watkins,16
2,Bournemouth,Eli Junior Kroupi,13
3,Brentford,Igor Thiago,22
4,Brighton,Danny Welbeck,13
5,Burnley,Zian Flemming,11
6,Chelsea,João Pedro,15
7,Crystal Palace,Jean-Philippe Mateta,12
8,Everton,Beto,9
9,Fulham,Harry Wilson,10


In [86]:
teams_export = pd.read_sql_query(
    "SELECT * FROM Teams",
    conn
)
teams_export.to_csv("Teams.csv", index=False)

In [87]:
players_export = pd.read_sql_query(
    "SELECT * FROM Players",
    conn
)
players_export.to_csv("Players.csv", index=False)

In [88]:
playerstats_export = pd.read_sql_query(
    "SELECT * FROM PlayerStats",
    conn
)
playerstats_export.to_csv("PlayerStats.csv", index=False)